Mathlib/Analysis/InnerProductSpace/TwoDim.lean

```diff
@@ -0,0 +1,676 @@
+#align orientation.kahler_mul Orientation.kahler_mul
+
+theorem normSq_kahler (x y : E) : Complex.normSq (o.kahler x y) = ‖x‖ ^ 2 * ‖y‖ ^ 2 := by
+  simpa [kahler_apply_apply, Complex.normSq, sq] using o.inner_sq_add_areaForm_sq x y
+#align orientation.norm_sq_kahler Orientation.normSq_kahler
+
+theorem abs_kahler (x y : E) : Complex.abs (o.kahler x y) = ‖x‖ * ‖y‖ := by
+  rw [← sq_eq_sq, Complex.sq_abs]
+  · linear_combination (norm := ring_nf) o.normSq_kahler x y; simp
+  · positivity
+  · positivity
+#align orientation.abs_kahler Orientation.abs_kahler
+
+theorem norm_kahler (x y : E) : ‖o.kahler x y‖ = ‖x‖ * ‖y‖ := by simpa using o.abs_kahler x y
+#align orientation.norm_kahler Orientation.norm_kahler
+
+theorem eq_zero_or_eq_zero_of_kahler_eq_zero {x y : E} (hx : o.kahler x y = 0) : x = 0 ∨ y = 0 := by
+  have : ‖x‖ * ‖y‖ = 0 := by simpa [hx] using (o.norm_kahler x y).symm
+  cases' eq_zero_or_eq_zero_of_mul_eq_zero this with h h
+  · left
+    simpa using h
+  · right
+    simpa using h
+#align orientation.eq_zero_or_eq_zero_of_kahler_eq_zero Orientation.eq_zero_or_eq_zero_of_kahler_eq_zero
+
+theorem kahler_eq_zero_iff (x y : E) : o.kahler x y = 0 ↔ x = 0 ∨ y = 0 := by
+  refine' ⟨o.eq_zero_or_eq_zero_of_kahler_eq_zero, _⟩
+  rintro (rfl | rfl) <;> simp
+#align orientation.kahler_eq_zero_iff Orientation.kahler_eq_zero_iff
+
+theorem kahler_ne_zero {x y : E} (hx : x ≠ 0) (hy : y ≠ 0) : o.kahler x y ≠ 0 := by
+  apply mt o.eq_zero_or_eq_zero_of_kahler_eq_zero
+  tauto
+#align orientation.kahler_ne_zero Orientation.kahler_ne_zero
+
+theorem kahler_ne_zero_iff (x y : E) : o.kahler x y ≠ 0 ↔ x ≠ 0 ∧ y ≠ 0 := by
+  refine' ⟨_, fun h => o.kahler_ne_zero h.1 h.2⟩
+  contrapose
+  simp only [not_and_or, Classical.not_not, kahler_apply_apply, Complex.real_smul]
+  rintro (rfl | rfl) <;> simp
+#align orientation.kahler_ne_zero_iff Orientation.kahler_ne_zero_iff
+
+theorem kahler_map {F : Type _} [NormedAddCommGroup F] [InnerProductSpace ℝ F]
+    [Fact (finrank ℝ F = 2)] (φ : E ≃ₗᵢ[ℝ] F) (x y : F) :
+    (Orientation.map (Fin 2) φ.toLinearEquiv o).kahler x y = o.kahler (φ.symm x) (φ.symm y) := by
+  simp [kahler_apply_apply, areaForm_map]
+#align orientation.kahler_map Orientation.kahler_map
+
+/-- The bilinear map `kahler` is invariant under pullback by a positively-oriented isometric
+automorphism. -/
+theorem kahler_comp_linearIsometryEquiv (φ : E ≃ₗᵢ[ℝ] E)
+    (hφ : 0 < LinearMap.det (φ.toLinearEquiv : E →ₗ[ℝ] E)) (x y : E) :
+    o.kahler (φ x) (φ y) = o.kahler x y :=
+  by simp [kahler_apply_apply, o.areaForm_comp_linearIsometryEquiv φ hφ]
+#align orientation.kahler_comp_linear_isometry_equiv Orientation.kahler_comp_linearIsometryEquiv
+
+end Orientation
+
+namespace Complex
+
+attribute [local instance] Complex.finrank_real_complex_fact
+
+@[simp]
+protected theorem areaForm (w z : ℂ) : Complex.orientation.areaForm w z = (conj w * z).im := by
+  let o := Complex.orientation
+  -- Porting note: split `simp only` for greater proof control
```

Comment:
Could you say why? Is there a `simp` regression here?

Reply:
The original proof was
```lean
protected theorem areaForm (w z : ℂ) : Complex.orientation.areaForm w z = (conj w * z).im := by
  let o := Complex.orientation
  simp only [o.areaForm_to_volumeForm, o.volumeForm_robust Complex.orthonormalBasisOneI rfl,
    Basis.det_apply, Matrix.det_fin_two, Basis.toMatrix_apply, toBasis_orthonormalBasisOneI,
    Matrix.cons_val_zero, coe_basisOneI_repr, Matrix.cons_val_one, Matrix.head_cons, mul_im,
    conj_re, conj_im]
  ring
```
which does not work, even after #4933. However putting brackets around two of the items in the `simp only` list makes it work:
```lean
protected theorem areaForm (w z : ℂ) : Complex.orientation.areaForm w z = (conj w * z).im := by
  let o := Complex.orientation
  simp only [o.areaForm_to_volumeForm, o.volumeForm_robust Complex.orthonormalBasisOneI rfl,
    (Basis.det_apply), Matrix.det_fin_two, (Basis.toMatrix_apply), toBasis_orthonormalBasisOneI,
    Matrix.cons_val_zero, coe_basisOneI_repr, Matrix.cons_val_one, Matrix.head_cons, mul_im,
    conj_re, conj_im]
  ring
```